File: sdk/go/market.go
package fyuz

import "context"

// DiscoverTab is a preset ordering for the discovery feed.
type DiscoverTab string

// The discovery feed orderings.
const (
	// TabTrending ranks by recent trading activity.
	TabTrending DiscoverTab = "trending"
	// TabNew ranks newest first.
	TabNew DiscoverTab = "new"
	// TabGainers ranks by 24h price change.
	TabGainers DiscoverTab = "gainers"
	// TabGraduating ranks by proximity to the $30,000 graduation threshold.
	TabGraduating DiscoverTab = "graduating"
)

// TokenStatus filters by bonding-curve state.
type TokenStatus string

const (
	// StatusBonding selects tokens still trading on the bonding curve, for
	// which this API is the only price source.
	StatusBonding TokenStatus = "bonding"
	// StatusGraduated selects tokens that have a PancakeSwap pair.
	StatusGraduated TokenStatus = "graduated"
)

// SortOrder is a sort direction.
type SortOrder string

const (
	// SortAsc sorts ascending.
	SortAsc SortOrder = "asc"
	// SortDesc sorts descending.
	SortDesc SortOrder = "desc"
)

// DiscoverOptions are the optional filters for Discover. The zero value asks
// for the API's defaults.
type DiscoverOptions struct {
	// Tab is a preset ordering. Empty leaves the server default.
	Tab DiscoverTab
	// Network is the chain slug, e.g. "bsc". Empty means all chains.
	Network string
	// MinMarketcap filters out tokens below this USD market cap.
	MinMarketcap *float64
	// MinVolume filters out tokens below this 24h USD volume.
	MinVolume *float64
	// MinHolders filters out tokens with fewer holders than this.
	MinHolders *int64
	// MaxAgeSecs keeps only tokens created within this many seconds.
	MaxAgeSecs *int64
	// Status filters by bonding-curve state.
	Status TokenStatus
	// Limit caps the number of rows. Defaults to 50 server-side.
	Limit *int64
}

func (o *DiscoverOptions) query() *query {
	q := newQuery()
	if o == nil {
		return q
	}
	q.str("tab", string(o.Tab))
	q.str("network", o.Network)
	q.float64p("minMarketcap", o.MinMarketcap)
	q.float64p("minVolume", o.MinVolume)
	q.int64p("minHolders", o.MinHolders)
	q.int64p("maxAgeSecs", o.MaxAgeSecs)
	q.str("status", string(o.Status))
	q.int64p("limit", o.Limit)
	return q
}

// ListTokensOptions are the optional filters and paging for ListTokens.
type ListTokensOptions struct {
	// OrderType is the field to sort by, e.g. "marketcap", "createdAt",
	// "score".
	OrderType string
	// OrderFlag is the sort direction, SortAsc or SortDesc.
	OrderFlag SortOrder
	// SearchWord free-text matches on name, symbol or address.
	SearchWord string
	// Network is the chain slug, e.g. "bsc".
	Network string
	// Style exact-matches the generated image style ("all" means no filter).
	Style string
	// IncludeNsfw includes tokens flagged not-safe-for-work.
	IncludeNsfw *bool
	// PageNumber is 1-based. Defaults to 1 server-side.
	PageNumber *int64
	// PageSize defaults to 20 server-side and is clamped to at most 100:
	// asking for more still returns 100 rows.
	PageSize *int64
}

func (o *ListTokensOptions) query() *query {
	q := newQuery()
	if o == nil {
		return q
	}
	q.str("orderType", o.OrderType)
	q.str("orderFlag", string(o.OrderFlag))
	q.str("searchWord", o.SearchWord)
	q.str("network", o.Network)
	q.str("style", o.Style)
	q.boolp("includeNsfw", o.IncludeNsfw)
	q.int64p("pageNumber", o.PageNumber)
	q.int64p("pageSize", o.PageSize)
	return q
}

// GetTokenOptions pages the trade list embedded in a token detail response.
type GetTokenOptions struct {
	// PageNumber is the 1-based page of trades. Defaults to 1 server-side.
	PageNumber *int64
	// PageSize is the trades per page. Defaults to 20 server-side.
	PageSize *int64
}

func (o *GetTokenOptions) query() *query {
	q := newQuery()
	if o == nil {
		return q
	}
	q.int64p("pageNumber", o.PageNumber)
	q.int64p("pageSize", o.PageSize)
	return q
}

// Health probes service liveness.
//
//	status, err := client.Health(ctx)
//	fmt.Println(status.Status) // "ok"
func (c *Client) Health(ctx context.Context) (*HealthStatus, error) {
	var out HealthStatus
	if err := c.get(ctx, "/health", nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// GetConfig returns the chain configuration the Fyuz frontend runs against.
//
//	cfg, err := client.GetConfig(ctx)
//	for _, chain := range cfg.Chains {
//		fmt.Println(chain.Name, chain.ChainID)
//	}
func (c *Client) GetConfig(ctx context.Context) (*ChainConfig, error) {
	var out ChainConfig
	if err := c.get(ctx, "/config", nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Discover returns the discovery feed: one row per token with market cap, 24h
// volume, buy/sell counts, holder count, price change and graduation progress.
//
// This is the highest-signal single endpoint and the recommended entry point
// for bulk market-data ingestion. Pass nil for the server defaults.
//
//	tokens, err := client.Discover(ctx, &fyuz.DiscoverOptions{
//		Tab:    fyuz.TabGraduating,
//		Status: fyuz.StatusBonding,
//		Limit:  fyuz.Int64(20),
//	})
func (c *Client) Discover(ctx context.Context, opts *DiscoverOptions) ([]DiscoverToken, error) {
	var out []DiscoverToken
	if err := c.get(ctx, "/discover", opts.query().encoded(), &out); err != nil {
		return nil, err
	}
	return out, nil
}

// ListTokens returns one page of the token list, with the full token record
// including bonding-curve virtual reserves.
//
// Use IterateTokens to walk every page automatically.
//
//	page, err := client.ListTokens(ctx, &fyuz.ListTokensOptions{
//		OrderType: "marketcap",
//		OrderFlag: fyuz.SortDesc,
//		PageSize:  fyuz.Int64(50),
//	})
//	fmt.Println(len(page.TokenList), "of", page.TokenCount)
func (c *Client) ListTokens(ctx context.Context, opts *ListTokensOptions) (*TokenPage, error) {
	var out TokenPage
	if err := c.get(ctx, "/tokens", opts.query().encoded(), &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// GetKing returns the current king of the hill: the highest-marketcap token
// still on the bonding curve.
//
// Returns (nil, nil) when the hill is empty. That is a real state, not an
// error — a token leaves the hill the moment it graduates, so between a
// graduation and the next launch there is no king. Check the pointer before
// dereferencing it.
//
//	king, err := client.GetKing(ctx)
//	if err != nil {
//		return err
//	}
//	if king == nil {
//		fmt.Println("no king right now")
//	}
//
// The API wraps the token in a {"king": ...} envelope; it is unwrapped here.
func (c *Client) GetKing(ctx context.Context) (*Token, error) {
	var out struct {
		King *Token `json:"king"`
	}
	if err := c.get(ctx, "/tokens/king", nil, &out); err != nil {
		return nil, err
	}
	return out.King, nil
}

// GetToken returns the full record for one token plus recent trades, holder
// distribution, the 15-minute price and one-day liquidity.
//
// network is the chain slug ("bsc"); tokenAddress is the 0x token address.
// Returns an error matching ErrNotFound when the token is unknown.
//
//	detail, err := client.GetToken(ctx, "bsc", "0x23a8...", nil)
//	if detail.TokenDetails.PairAddress == nil {
//		// still on the bonding curve: no DEX pool exists for this token
//	}
func (c *Client) GetToken(ctx context.Context, network, tokenAddress string, opts *GetTokenOptions) (*TokenDetail, error) {
	var out TokenDetail
	path := "/tokens/" + escapePath(network) + "/" + escapePath(tokenAddress)
	if err := c.get(ctx, path, opts.query().encoded(), &out); err != nil {
		return nil, err
	}
	return &out, nil
}

File: sdk/rust/src/api/market.rs
//! Market data: service metadata, tokens, trades, candles and analytics.

use serde_json::json;

use crate::client::{seg, FyuzClient, QueryPairs};
use crate::error::Result;
use crate::models::{
    Candle, ChainConfig, DiscoverToken, HealthStatus, KingResponse, RecentTrades, Token,
    TokenAnalytics, TokenDetail, TokenPage, TopTrader, Trade,
};
use crate::pagination::TokenPager;
use crate::params::{
    ChartDataParams, DiscoverParams, ListTokensParams, RecentTradesParams, ToQuery,
    TokenDetailParams, TokenTradesParams,
};

impl FyuzClient {
    /// `GET /health` — liveness probe.
    ///
    /// ```no_run
    /// # async fn demo() -> Result<(), fyuz::Error> {
    /// let client = fyuz::FyuzClient::new()?;
    /// assert_eq!(client.health().await?.status, "ok");
    /// # Ok(())
    /// # }
    /// ```
    pub async fn health(&self) -> Result<HealthStatus> {
        self.get_json("/health", QueryPairs::new()).await
    }

    /// `GET /config` — the chains this deployment serves.
    ///
    /// ```no_run
    /// # async fn demo() -> Result<(), fyuz::Error> {
    /// let client = fyuz::FyuzClient::new()?;
    /// for chain in client.get_config().await?.chains {
    ///     println!("{} (chain id {})", chain.name, chain.chain_id);
    /// }
    /// # Ok(())
    /// # }
    /// ```
    pub async fn get_config(&self) -> Result<ChainConfig> {
        self.get_json("/config", QueryPairs::new()).await
    }

    /// `GET /discover` — the discovery feed, and the best single endpoint for
    /// bulk market-data ingestion.
    ///
    /// One row per token with market cap, 24h volume, buy/sell counts, holder
    /// count, price change and graduation progress.
    ///
    /// ```no_run
    /// # async fn demo() -> Result<(), fyuz::Error> {
    /// use fyuz::{DiscoverParams, DiscoverTab};
    ///
    /// let client = fyuz::FyuzClient::new()?;
    /// let almost_there = client
    ///     .discover(&DiscoverParams::new().tab(DiscoverTab::Graduating).limit(20))
    ///     .await?;
    ///
    /// for token in &almost_there {
    ///     println!("{:<8} {:>6.2}% of the way to a PancakeSwap pair", token.symbol, token.graduation_pct);
    /// }
    /// # Ok(())
    /// # }
    /// ```
    pub async fn discover(&self, params: &DiscoverParams) -> Result<Vec<DiscoverToken>> {
        self.get_json("/discover", params.to_query()).await
    }

    /// `GET /tokens` — one page of the token list.
    ///
    /// Use [`token_pages`](Self::token_pages) to walk every page instead of
    /// paging by hand.
    ///
    /// ```no_run
    /// # async fn demo() -> Result<(), fyuz::Error> {
    /// use fyuz::{ListTokensParams, SortDirection};
    ///
    /// let client = fyuz::FyuzClient::new()?;
    /// let page = client
    ///     .list_tokens(
    ///         &ListTokensParams::new()
    ///             .order_type("marketcap")
    ///             .order_flag(SortDirection::Desc)
    ///             .page_size(50),
    ///     )
    ///     .await?;
    ///
    /// println!("{} of {} tokens", page.token_list.len(), page.token_count);
    /// # Ok(())
    /// # }
    /// ```
    pub async fn list_tokens(&self, params: &ListTokensParams) -> Result<TokenPage> {
        self.get_json("/tokens", params.to_query()).await
    }

    /// Walk every page of `GET /tokens`.
    ///
    /// The returned [`TokenPager`] issues one request per page and stops when
    /// the server's `tokenCount` is covered.
    ///
    /// ```no_run
    /// # async fn demo() -> Result<(), fyuz::Error> {
    /// use fyuz::ListTokensParams;
    ///
    /// let client = fyuz::FyuzClient::new()?;
    /// let mut pager = client.token_pages(&ListTokensParams::new().page_size(100));
    ///
    /// while let Some(token) = pager.next_token().await? {
    ///     println!("{} {}", token.token_symbol, token.marketcap);
    /// }
    /// # Ok(())
    /// # }
    /// ```
    pub fn token_pages(&self, params: &ListTokensParams) -> TokenPager {
        TokenPager::new(self.clone(), params.clone())
    }

    /// `GET /tokens/king` — the current king of the hill: the
    /// highest-market-cap token still on the bonding curve.
    ///
    /// `None` means the hill is empty, which is a real state rather than an
    /// error: a token leaves the moment it graduates, so between a graduation
    /// and the next launch there is no king.
    ///
    /// The API wraps the token in a `{"king": …}` envelope; it is unwrapped
    /// here.
    ///
    /// ```no_run
    /// # async fn demo() -> Result<(), fyuz::Error> {
    /// let client = fyuz::FyuzClient::new()?;
    /// match client.get_king().await? {
    ///     Some(king) => println!("{} at ${}", king.token_name, king.marketcap),
    ///     None => println!("no king right now"),
    /// }
    /// # Ok(())
    /// # }
    /// ```
    pub async fn get_king(&self) -> Result<Option<Token>> {
        let body: KingResponse = self.get_json("/tokens/king", QueryPairs::new()).await?;
        Ok(body.king)
    }

    /// `GET /tokens/{network}/{tokenAddress}` — full token record plus recent
    /// trades, holders, 15-minute price and 1-day liquidity.
    ///
    /// `curve_holding` is `None` when the on-chain read failed. That means
    /// **unknown**, not zero — do not render it as `0`.
    ///
    /// Answers `404` for an unknown token; check with
    /// [`Error::is_not_found`](crate::Error::is_not_found).
    ///
    /// ```no_run
    /// # async fn demo() -> Result<(), fyuz::Error> {
    /// use fyuz::TokenDetailParams;
    ///
    /// let client = fyuz::FyuzClient::new()?;
    /// let detail = client
    ///     .get_token("bsc", "0x23a84ba5bf7bf3236491fa2b5a9807a274337135", &TokenDetailParams::new())
    ///     .await?;
    ///
    /// let graduated = detail.token_details.pair_address.is_some();
    /// println!("graduated: {graduated}, {} trades", detail.trades_count);
    /// # Ok(())
    /// # }
    /// ```
    pub async fn get_token(
        &self,
        network: &str,
        token_address: &str,
        params: &TokenDetailParams,
    ) -> Result<TokenDetail> {
        self.get_json(
            &format!("/tokens/{}/{}", seg(network), seg(token_address)),
            params.to_query(),
        )
            .await
    }

    /// `GET /trades/recent` — the platform-wide trade feed.
    ///
    /// Pass the highest [`Trade::id`] you have already seen as
    /// [`RecentTradesParams::latest_trade_id`] to poll incrementally.
    ///
    /// ```no_run
    /// # async fn demo() -> Result<(), fyuz::Error> {
    /// use fyuz::RecentTradesParams;
    ///
    /// let client = fyuz::FyuzClient::new()?;
    /// let mut cursor = 0i64;
    ///
    /// let feed = client
    ///     .get_recent_trades(&RecentTradesParams::new().latest_trade_id(cursor))
    ///     .await?;
    ///
    /// if let Some(newest) = feed.trades.first() {
    ///     cursor = newest.id;
    /// }
    /// # let _ = cursor;
    /// # Ok(())
    /// # }
    /// ```
    pub async fn get_recent_trades(&self, params: &RecentTradesParams) -> Result<RecentTrades> {
        self.get_json("/trades/recent", params.to_query()).await
    }

    /// `POST /trades` — one token's trade history, newest first.
    ///
    /// This is a `POST` only because the token address travels in the body; it
    /// is a read, it changes nothing, and it needs no authentication.
    ///
    /// ```no_run
    /// # async fn demo() -> Result<(), fyuz::Error> {
    /// use fyuz::TokenTradesParams;
    ///
    /// let client = fyuz::FyuzClient::new()?;
    /// let trades = client
    ///     .get_token_trades(
    ///         "0x23a84ba5bf7bf3236491fa2b5a9807a274337135",
    ///         &TokenTradesParams::new().limit(100),
    ///     )
    ///     .await?;
    ///
    /// for trade in &trades {
    ///     // eth_amount is a decimal string: keep it as one.
    ///     println!("{} {} BNB", trade.trade_side, trade.eth_amount);
    /// }
    /// # Ok(())
    /// # }
    /// ```
    pub async fn get_token_trades(
        &self,
        token_address: &str,
        params: &TokenTradesParams,
    ) -> Result<Vec<Trade>> {
        self.post_json(
            "/trades",
            params.to_query(),
            json!({ "tokenAddress": token_address }),
        )
        .await
    }

    /// `GET /trades/getChartData` — OHLCV candles, shaped for TradingView-style
    /// charting libraries.
    ///
    /// ```no_run
    /// # async fn demo() -> Result<(), fyuz::Error> {
    /// use fyuz::ChartDataParams;
    ///
    /// let client = fyuz::FyuzClient::new()?;
    /// let candles = client
    ///     .get_chart_data(&ChartDataParams::new(
    ///         "0x23a84ba5bf7bf3236491fa2b5a9807a274337135",
    ///         "5",
    ///         1_750_000_000,
    ///         1_750_086_400,
    ///     ))
    ///     .await?;
    ///
    /// println!("{} candles", candles.len());
    /// # Ok(())
    /// # }
    /// ```
    pub async fn get_chart_data(&self, params: &ChartDataParams) -> Result<Vec<Candle>> {
        self.get_json("/trades/getChartData", params.to_query()).await
    }

    /// `GET /analytics/token/{network}/{address}` — holder concentration,
    /// buy/sell split, volume, graduation progress and bundle-buyer detection.
    ///
    /// Answers `404` for an unknown token.
    ///
    /// ```no_run
    /// # async fn demo() -> Result<(), fyuz::Error> {
    /// let client = fyuz::FyuzClient::new()?;
    /// let a = client
    ///     .get_token_analytics("bsc", "0x23a84ba5bf7bf3236491fa2b5a9807a274337135")
    ///     .await?;
    ///
    /// if a.bundle_flag {
    ///     println!("{} addresses bought in the launch block", a.bundle_buyers);
    /// }
    /// # Ok(())
    /// # }
    /// ```
    pub async fn get_token_analytics(
        &self,
        network: &str,
        address: &str,
    ) -> Result<TokenAnalytics> {
        self.get_json(
            &format!("/analytics/token/{}/{}", seg(network), seg(address)),
            QueryPairs::new(),
        )
        .await
    }

    /// `GET /analytics/top-traders/{network}/{address}` — a token's most
    /// profitable traders, with the creator flagged.
    ///
    /// ```no_run
    /// # async fn demo() -> Result<(), fyuz::Error> {
    /// let client = fyuz::FyuzClient::new()?;
    /// let traders = client
    ///     .get_top_traders("bsc", "0x23a84ba5bf7bf3236491fa2b5a9807a274337135")
    ///     .await?;
    ///
    /// for t in traders.iter().take(5) {
    ///     println!("{} {:+.2} USD", t.address, t.total_pnl_usd);
    /// }
    /// # Ok(())
    /// # }
    /// ```
    pub async fn get_top_traders(&self, network: &str, address: &str) -> Result<Vec<TopTrader>> {
        self.get_json(
            &format!("/analytics/top-traders/{}/{}", seg(network), seg(address)),
            QueryPairs::new(),
        )
        .await
    }
}
